src/ray/core_worker/task_manager.cc

```diff
@@ -358,6 +361,12 @@ std::optional<rpc::ErrorType> TaskManager::ResubmitTask(
 
   UpdateReferencesForResubmit(spec, task_deps);
 
+  // TODO(can-anyscale): There is an existing race condition (even before this comment
+  // was added) where a task can still be retried after its retry count has reached
+  // zero—for example, in the case of task cancellation—because this code is still
+  // reachable. The root cause is that we cannot guard the callback invocation with
+  // the TaskManager mutex, since the callback itself acquires that mutex. As a result,
+  // we cannot guarantee the task's state before the callback runs.
```

Comment:
TODOs should have linked issues that provide additional context. We can make this a little more concise e.g.
```suggestion
  // TODO(can-anyscale): There is a race condition where a task can still be retried 
  // after its retry count has reached zero. Additional information in <ISSUE>
```